aws/logs_monitoring/lambda_function.py

```diff
@@ -53,86 +63,217 @@
 # Strip any trailing and leading whitespace from the API key
 DD_API_KEY = DD_API_KEY.strip()
 
-cloudtrail_regex = re.compile(
-    "\d+_CloudTrail_\w{2}-\w{4,9}-\d_\d{8}T\d{4}Z.+.json.gz$", re.I
-)
-ip_regex = re.compile("\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}", re.I)
 DD_SOURCE = "ddsource"
 DD_CUSTOM_TAGS = "ddtags"
 DD_SERVICE = "service"
 DD_HOST = "host"
-DD_FORWARDER_VERSION = "1.2.2"
+DD_FORWARDER_VERSION = "1.3.0"
 
 # Pass custom tags as environment variable, ensure comma separated, no trailing comma in envvar!
 DD_TAGS = os.environ.get("DD_TAGS", "")
 
 
-class DatadogConnection(object):
-    def __init__(self, host, port, ddApiKey):
+class RetriableException(Exception):
+    pass
+
+
+class ScrubbingException(Exception):
+    pass
+
+
+class DatadogClient(object):
+    """
+    Client that implements a linear retrying logic to send a batch of logs.
+    """
+
+    def __init__(self, client, max_retries=3, backoff=1):
+        self._client = client
+        self._max_retries = max_retries
+        self._backoff = backoff
+
+    def send(self, logs):
+        retry = 0
+        while retry <= self._max_retries:
+            if retry > 0:
+                time.sleep(self._backoff)
+            try:
+                self._client.send(logs)
+            except RetriableException as e:
+                retry += 1
+                continue
+            return
+        raise Exception("max number of retries reached: {}".format(self._max_retries))
+
+    def __enter__(self):
+        self._client.__enter__()
+        return self
+
+    def __exit__(self, ex_type, ex_value, traceback):
+        self._client.__exit__(ex_type, ex_value, traceback)
+
+
+class DatadogTCPClient(object):
+    """
+    Client that sends a batch of logs over TCP.
+    """
+
+    def __init__(self, host, port, api_key, scrubber):
         self.host = host
         self.port = port
-        self.api_key = ddApiKey
+        self._api_key = api_key
+        self._scrubber = scrubber
         self._sock = None
 
     def _connect(self):
-        s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
-        s = ssl.wrap_socket(s)
-        s.connect((self.host, self.port))
-        return s
+        sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        sock = ssl.wrap_socket(sock)
+        sock.connect((self.host, self.port))
```

Comment:
connect can fail with an exception, don't we want to retry also in case of connection failure? also it's be nice to specify a timeout